flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/serializer/data/writer/BinaryWriter.java

```diff
@@ -98,9 +105,21 @@ static void write(
                 writer.writeShort(pos, (short) o);
                 break;
             case INTEGER:
+                writer.writeInt(pos, (int) o);
+                break;
             case DATE:
+                if (o instanceof DateData) {
+                    writer.writeDate(pos, (DateData) o);
+                } else {
+                    writer.writeInt(pos, (int) o);
+                }
+                break;
```

Comment:
We may add some comments here, clarifying that the `writeInt` paths are kept for compatibility.